### ISSUE TITLE
Change *-empty-line-before keywords behaviour to ignore shared-line comments

### ISSUE THREAD

**Issue by hudochenkov** [stylelint/stylelint#2237]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

This issue is similar to #2212. I want to normalize behaviour with shared-line comments and ignore them for more options. If previous node is shared-line comment, than we should check node before shared-line comment.

> Which rule, if any, is this issue related to?

`at-rule-empty-line-before`
except: ["blockless-after-same-name-blockless", "blockless-group"]
ignore: ["blockless-after-same-name-blockless"] ("blockless-group" for `ignore` has very different criteria than for `except`. #2240)

`custom-property-empty-line-before`
except: ["after-custom-property"]

`declaration-empty-line-before`
except: ["after-declaration"]
ignore: ["after-declaration"]

`rule-nested-empty-line-before`
except: ["after-rule"]

> What CSS is needed to reproduce this issue?

```css
a {

    color: #0098db; /* sfs */
    text-decoration: none;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "declaration-empty-line-before": ["always", {
        "except": ["after-declaration"]
    }],
  }
}
```

> Which version of stylelint are you using?

e.g. `7.7.1`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warning was flagged:

```
4:5 error Expected empty line before declaration (declaration-empty-line-before)
```

**Comment by jeddy3** [comment]
@hudochenkov It's really appreciated that you're doing this audit! 

> I want to normalize behaviour with shared-line comments and ignore them for more options. If previous node is shared-line comment, than we should check node before shared-line comment.

Agreed. This is likely an oversight which touches on quite a few rules and options.

As with #2212, these will be breaking changes and as such will need to go into the `v8` branch.

**Comment by hudochenkov** [comment]
@jeddy3 #2240, #2239, #2212, and this issue are in one place kind of. I'm thinking about two options to manage them:

1. Make one PR with all changes.
2. Close #2240, #2239 first in the first PR. Then close #2212, and this issue in the second PR, after first one will be accepted.

What option should I choose?

**Comment by jeddy3** [comment]
I think option one might be more productive for you and I don't think will add much to the reviewing burden.

**Comment by jeddy3** [comment]
I think waiting on https://github.com/stylelint/stylelint/issues/1663 being merged into `master` might be prudent though, as I think we're setting ourselves up for a lot of merge conflicts.

**Comment by hudochenkov** [comment]
@jeddy3 got it! Ping me, please, when it's ready, so I can work on this PR.

**Comment by jeddy3** [comment]
Will do. We're [working through](https://github.com/stylelint/stylelint/issues/2079#issue-190470316) some big project scope/direction decisions for `8.0.0` at the moment, but there is light at the end of the tunnel. So, hopefully, the deprecations will be decided upon soon.

**Comment by davidtheclark** [comment]
Closed by #2262.
